File: Homework_7/Homework7_ex1/Homework7_ex1/Form1.vb
Imports System.Reflection.Emit
Imports System.Windows.Forms.VisualStyles.VisualStyleElement

Public Class Form1

    Public b As Bitmap
    Public g As Graphics
    Public r As New Random
    Public PenTrajectory As New Pen(Color.OrangeRed, 2)
    Public b2 As Bitmap
    Public g2 As Graphics
    Public b3 As Bitmap
    Public g3 As Graphics

    Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click

        Me.b = New Bitmap(Me.PictureBox1.Width, Me.PictureBox1.Height)
        Me.g = Graphics.FromImage(b)
        Me.g.SmoothingMode = Drawing2D.SmoothingMode.HighQuality
        Me.g.Clear(Color.White)
        Me.b2 = New Bitmap(Me.PictureBox2.Width, Me.PictureBox2.Height)
        Me.g2 = Graphics.FromImage(b2)
        Me.g2.SmoothingMode = Drawing2D.SmoothingMode.HighQuality
        Me.g2.Clear(Color.White)

        Dim TrialsCount As Integer = TrackBar1.Value
        Dim NumerOfTrajectories As Integer = TrackBar2.Value
        Dim SuccessProbability As Double = TrackBar3.Value / TrialsCount

        Dim minX As Double = 0
        Dim maxX As Double = TrialsCount
        Dim minY As Double = 0
        Dim maxY As Double = TrialsCount

        Dim interarrival = 0

        Dim interarrivalDistribution As New Dictionary(Of Integer, Integer)

        Me.b3 = New Bitmap(Me.PictureBox3.Width, Me.PictureBox3.Height)
        Me.g3 = Graphics.FromImage(b3)
        Me.g3.SmoothingMode = Drawing2D.SmoothingMode.HighQuality
        Me.g3.Clear(Color.White)

        g3.DrawRectangle(New Pen(Color.Black, 1), 0, 0, PictureBox3.Width - 1, PictureBox3.Height - 1)

        Dim VirtualWindow As New Rectangle(20, 20, Me.b.Width - 40, Me.b.Height - 40)

        g.DrawRectangle(Pens.DarkSlateGray, VirtualWindow)

        Dim alt(Me.PictureBox1.Height) As Integer
        For i As Integer = 0 To Me.PictureBox1.Height
            alt(i) = 0
        Next

        For i As Integer = 1 To NumerOfTrajectories

            Dim Punti As New List(Of Point)
            Dim Y As Double = 0
            For X As Integer = 1 To TrialsCount
                Dim Uniform As Double = r.NextDouble
                If Uniform < SuccessProbability Then
                    Y = Y + 1
                    If interarrival <> 0 Then
                        If Not interarrivalDistribution.ContainsKey(interarrival) Then
                            interarrivalDistribution.Add(interarrival, 1)
                        Else
                            interarrivalDistribution.Item(interarrival) = interarrivalDistribution.Item(interarrival) + 1
                        End If
                        interarrival = 0
                    End If
                Else
                    interarrival = interarrival + 1
                End If
                Dim xDevice As Integer = FromXRealToXVirtual(X, minX, maxX, VirtualWindow.Left, VirtualWindow.Width)
                Dim YDevice As Integer = FromYRealToYVirtual(Y, minY, maxY, VirtualWindow.Top, VirtualWindow.Height)
                Punti.Add(New Point(xDevice, YDevice))
            Next
            g.DrawLines(PenTrajectory, Punti.ToArray)
            alt(Punti.Last.Y) += 60
        Next

        Me.PictureBox1.Image = b
        Dim pen As New Pen(Color.Black, 5)
        Me.PictureBox2.Image = b2

        For i As Integer = 0 To alt.Length - 1

            g2.DrawLine(pen, 0, i, alt(i), i)

        Next

        drawVerticalChart(b3, g3, PictureBox3, interarrivalDistribution, TrialsCount)
        PictureBox3.Image = b3

    End Sub

    Function drawVerticalChart(b As Bitmap, g As Graphics, p As PictureBox, distr As Dictionary(Of Integer, Integer), n As Integer)
        Dim j As Integer
        j = 0
        Dim s As Integer
        s = p.Width / distr.Count

        g.DrawRectangle(New Pen(Color.Black), 0, 0, p.Width - 1, p.Height - 1)

        For Each pair As KeyValuePair(Of Integer, Integer) In distr
            Dim virtualX As Double = FromXRealToXVirtual(pair.Value, 0, n, 0, p.Height)
            Dim r As Rectangle = New Rectangle(j + 1, p.Height - CType(virtualX, Integer) - 1, s, CType(virtualX, Integer))
            g.DrawRectangle(New Pen(Color.Black), j + 1, p.Height - CType(virtualX, Integer) - 1, s, CType(virtualX, Integer))
            g.FillRectangle(New SolidBrush(Color.Orange), r)
            j = j + s

        Next

        p.Image = b

    End Function

    Function FromXRealToXVirtual(X As Double,
                                 minX As Double, maxX As Double,
                                 Left As Integer, W As Integer) As Integer

        If (maxX - minX) = 0 Then
            Return 0
        End If

        Return Left + W * (X - minX) / (maxX - minX)

    End Function

    Function FromYRealToYVirtual(Y As Double,
                                minY As Double, maxY As Double,
                                Top As Integer, H As Integer) As Integer

        If (maxY - minY) = 0 Then
            Return 0
        End If

        Return Top + H - H * (Y - minY) / (maxY - minY)

    End Function



    Private Sub Label1_Click(sender As Object, e As EventArgs) Handles Label1.Click

    End Sub

    Private Sub TrackBar1_Scroll(sender As Object, e As EventArgs) Handles TrackBar1.Scroll
        Label1.Text = "Number of trials: " + TrackBar1.Value.ToString
    End Sub

    Private Sub TrackBar2_Scroll(sender As Object, e As EventArgs) Handles TrackBar2.Scroll
        Label2.Text = "Number of trajectories: " + TrackBar2.Value.ToString
    End Sub

    Private Sub TrackBar3_Scroll(sender As Object, e As EventArgs) Handles TrackBar3.Scroll
        Label3.Text = "Value of λ: " + (TrackBar3.Value).ToString
    End Sub

    Private Sub Button2_Click(sender As Object, e As EventArgs) Handles Button2.Click

        Me.b = New Bitmap(Me.PictureBox1.Width, Me.PictureBox1.Height)
        Me.g = Graphics.FromImage(b)
        Me.g.SmoothingMode = Drawing2D.SmoothingMode.HighQuality
        Me.g.Clear(Color.White)
        Me.b2 = New Bitmap(Me.PictureBox2.Width, Me.PictureBox2.Height)
        Me.g2 = Graphics.FromImage(b2)
        Me.g2.SmoothingMode = Drawing2D.SmoothingMode.HighQuality
        Me.g2.Clear(Color.White)

        Dim TrialsCount As Integer = TrackBar1.Value
        Dim NumerOfTrajectories As Integer = TrackBar2.Value
        Dim SuccessProbability As Double = TrackBar3.Value / TrialsCount

        Dim minX As Double = 0
        Dim maxX As Double = TrialsCount
        Dim minY As Double = 0
        Dim maxY As Double = TrialsCount

        Dim interarrival = 0

        Dim interarrivalDistribution As New Dictionary(Of Integer, Integer)

        Me.b3 = New Bitmap(Me.PictureBox3.Width, Me.PictureBox3.Height)
        Me.g3 = Graphics.FromImage(b3)
        Me.g3.SmoothingMode = Drawing2D.SmoothingMode.HighQuality
        Me.g3.Clear(Color.White)


        Dim VirtualWindow As New Rectangle(20, 20, Me.b.Width - 40, Me.b.Height - 40)

        g.DrawRectangle(Pens.DarkSlateGray, VirtualWindow)

        Dim alt(Me.PictureBox1.Height) As Integer
        For i As Integer = 0 To Me.PictureBox1.Height
            alt(i) = 0
        Next

        For i As Integer = 1 To NumerOfTrajectories

            Dim Punti As New List(Of Point)
            Dim Y As Double = 0
            For X As Integer = 1 To TrialsCount
                Dim Uniform As Double = r.NextDouble
                If Uniform <SuccessProbability Then
                    Y = Y + 1
                    If interarrival <> 0 Then
                        If Not interarrivalDistribution.ContainsKey(interarrival) Then
                            interarrivalDistribution.Add(interarrival, 1)
                        Else
                            interarrivalDistribution.Item(interarrival) = interarrivalDistribution.Item(interarrival) + 1
                        End If
                        interarrival = 0
                    End If
                Else
                    interarrival = interarrival + 1
                End If
                Dim xDevice As Integer = FromXRealToXVirtual(X, minX, maxX, VirtualWindow.Left, VirtualWindow.Width)
                Dim YDevice As Integer = FromYRealToYVirtual(Y * TrialsCount / (X + 1), minY, maxY, VirtualWindow.Top, VirtualWindow.Height)
                Punti.Add(New Point(xDevice, YDevice))
            Next
            g.DrawLines(PenTrajectory, Punti.ToArray)
            alt(Punti.Last.Y) += 60
        Next

        Me.PictureBox1.Image = b
        Dim pen As New Pen(Color.Black, 4)
        Me.PictureBox2.Image = b2
        For i As Integer = 0 To alt.Length - 1

            g2.DrawLine(pen, 0, i, alt(i), i)

        Next

        drawVerticalChart(b3, g3, PictureBox3, interarrivalDistribution, TrialsCount)
        PictureBox3.Image = b3

    End Sub

    Private Sub Button3_Click(sender As Object, e As EventArgs) Handles Button3.Click

        Me.b = New Bitmap(Me.PictureBox1.Width, Me.PictureBox1.Height)
        Me.g = Graphics.FromImage(b)
        Me.g.SmoothingMode = Drawing2D.SmoothingMode.HighQuality
        Me.g.Clear(Color.White)
        Me.b2 = New Bitmap(Me.PictureBox2.Width, Me.PictureBox2.Height)
        Me.g2 = Graphics.FromImage(b2)
        Me.g2.SmoothingMode = Drawing2D.SmoothingMode.HighQuality
        Me.g2.Clear(Color.White)

        Dim TrialsCount As Integer = TrackBar1.Value
        Dim NumerOfTrajectories As Integer = TrackBar2.Value
        Dim SuccessProbability As Double = TrackBar3.Value / TrialsCount

        Dim minX As Double = 0
        Dim maxX As Double = TrialsCount
        Dim minY As Double = 0
        Dim maxY As Double = TrialsCount

        Dim interarrival = 0

        Dim interarrivalDistribution As New Dictionary(Of Integer, Integer)

        Me.b3 = New Bitmap(Me.PictureBox3.Width, Me.PictureBox3.Height)
        Me.g3 = Graphics.FromImage(b3)
        Me.g3.SmoothingMode = Drawing2D.SmoothingMode.HighQuality
        Me.g3.Clear(Color.White)

        Dim VirtualWindow As New Rectangle(20, 20, Me.b.Width - 40, Me.b.Height - 40)

        g.DrawRectangle(Pens.DarkSlateGray, VirtualWindow)

        Dim alt(Me.PictureBox1.Height) As Integer
        For i As Integer = 0 To Me.PictureBox1.Height
            alt(i) = 0
        Next

        For i As Integer = 1 To NumerOfTrajectories

            Dim Punti As New List(Of Point)
            Dim Y As Double = 0
            For X As Integer = 1 To TrialsCount
                Dim Uniform As Double = r.NextDouble
                If Uniform < SuccessProbability Then
                    Y = Y + 1
                    If interarrival <> 0 Then
                        If Not interarrivalDistribution.ContainsKey(interarrival) Then
                            interarrivalDistribution.Add(interarrival, 1)
                        Else
                            interarrivalDistribution.Item(interarrival) = interarrivalDistribution.Item(interarrival) + 1
                        End If
                        interarrival = 0
                    End If
                Else
                    interarrival = interarrival + 1
                End If
                Dim xDevice As Integer = FromXRealToXVirtual(X, minX, maxX, VirtualWindow.Left, VirtualWindow.Width)
                Dim YDevice As Integer = FromYRealToYVirtual(Y * Math.Sqrt(TrialsCount) / Math.Sqrt(X + 1), minY, maxY, VirtualWindow.Top, VirtualWindow.Height)
                Punti.Add(New Point(xDevice, YDevice))
            Next
            g.DrawLines(PenTrajectory, Punti.ToArray)
            alt(Punti.Last.Y) += 60
        Next

        Me.PictureBox1.Image = b
        Dim pen As New Pen(Color.Black, 4)
        Me.PictureBox2.Image = b2
        For i As Integer = 0 To alt.Length - 1

            g2.DrawLine(pen, 0, i, alt(i), i)

        Next

        drawVerticalChart(b3, g3, PictureBox3, interarrivalDistribution, TrialsCount)
        PictureBox3.Image = b3

    End Sub

End Class

File: Homework_8/Hmw8_1/Hmw8_1/Form1.cs
using System.Reflection;
using System;

namespace Hmw8_1
{
    public partial class Form1 : Form
    {

        Bitmap b, b2, b3;
        Graphics g, g2, g3;


        Rectangle rect1;
        Pen PenTrajectory;
        Random r;

        double minX;
        double maxX;
        double minY;
        double maxY;
        int numberOfPoints = 1000;


        List<Point> points;

        public Form1()
        {
            InitializeComponent();

            b = new Bitmap(pictureBox1.Width, pictureBox1.Height);
            g = Graphics.FromImage(b);

            b2 = new Bitmap(pictureBox2.Width, pictureBox2.Height);
            g2 = Graphics.FromImage(b2);

            b3 = new Bitmap(pictureBox3.Width, pictureBox3.Height);
            g3 = Graphics.FromImage(b3);

            this.b = new Bitmap(this.pictureBox1.Width, this.pictureBox1.Height);
            this.b2 = new Bitmap(this.pictureBox2.Width, this.pictureBox2.Height);
            this.b3 = new Bitmap(this.pictureBox3.Width, this.pictureBox3.Height);
            this.r = new Random();
        }

        public int FromXRealToXVirtual(double X, double minX, double maxX, int L, int W)
        {
            return L + (int)(W * ((X - minX) / (maxX - minX)));
        }

        private void trackBar1_Scroll(object sender, EventArgs e)
        {
            numberOfPoints = trackBar1.Value;
            label1.Text = "# points : " + numberOfPoints.ToString();
        }

        public int FromYRealToYVirtual(double Y, double minY, double maxY, int T, int H)
        {
            return T + (int)(H - H * ((Y - minY) / (maxY - minY)));
        }

        private void button1_Click(object sender, EventArgs e)
        {
            this.g = Graphics.FromImage(this.b);
            this.g2 = Graphics.FromImage(this.b2);
            this.g3 = Graphics.FromImage(this.b3);


            this.g.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality;
            this.g.Clear(Color.White);

            this.g2.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality;
            this.g2.Clear(Color.White);

            this.g3.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality;
            this.g3.Clear(Color.White);

            points = new List<Point>();


            minX = -100d;
            maxX = 100d;
            minY = -100d;
            maxY = 100d;

            rect1 = new Rectangle(20, 20, this.b.Width - 40, this.b.Height - 40);
            g.DrawRectangle(Pens.Black, rect1);

            Random module = new Random();
            Random angle = new Random();
            Dictionary<int, int> xDistr = new Dictionary<int, int>();
            Dictionary<int, int> yDistr = new Dictionary<int, int>();

            int radius = 100;

            for (int i = 0; i < numberOfPoints; i++)
            {
                double p_rand = module.NextDouble() * radius;
                double p_angle = angle.NextDouble() * 2 * Math.PI;
                double x = p_rand * Math.Cos(p_angle);
                double y = p_rand * Math.Sin(p_angle);

                Point p = new Point(FromXRealToXVirtual(x, minX, maxX, rect1.Left, rect1.Width), FromYRealToYVirtual(y, minY, maxY, rect1.Top, rect1.Height));
                points.Add(p);

                if (xDistr.ContainsKey(p.X))
                    xDistr[p.X]++;
                else
                    xDistr.Add(p.X, 1);

                if (yDistr.ContainsKey(p.Y))
                    yDistr[p.Y]++;
                else
                    yDistr.Add(p.Y, 1);
            }

            foreach (Point p in points)
            {
                Rectangle rect = new Rectangle(p.X - 1, p.Y - 1, 2, 2);
                g.FillEllipse(Brushes.Black, rect);
            }

            Rectangle hor_Histo = new Rectangle(20, 20, this.b3.Width - 40, this.b3.Height - 40);
            g3.DrawRectangle(Pens.Black, hor_Histo);
            createIstogramHoriz(hor_Histo, g3, 20, this.b3.Width - 40, yDistr);


            Rectangle vert_Histo = new Rectangle(20, 20, this.b2.Width - 40, this.b2.Height - 40);
            g2.DrawRectangle(Pens.Black, vert_Histo);
            createIstogramVert(vert_Histo, g2, 20 + this.b2.Height - 40, this.b2.Height - 40, xDistr);

            pictureBox2.Image = b2;
            pictureBox3.Image = b3;
            pictureBox1.Image = b;
        }

        public void createIstogramHoriz(Rectangle istogramSpace, Graphics g, int x, int w, Dictionary<int, int> distances)
        {
            int max_value = 0;
            foreach (int key in distances.Keys)
            {
                if (distances[key] > max_value)
                    max_value = distances[key];
            }

            Pen istoPen = new Pen(Color.Orange, 2);

            foreach (int key in distances.Keys)
            {
                double currentInterValue = distances[key];
                double pct = (double)distances[key] / (double)max_value;
                g.DrawLine(istoPen,
                           new PointF(x, key),
                           new PointF(x + ((int)(pct * w)), key)
                );
            }

        }

        public void createIstogramVert(Rectangle istogramSpace, Graphics g, int y, int w, Dictionary<int, int> distances)
        {
            int max_value = 0;
            foreach (int key in distances.Keys)
            {
                if (distances[key] > max_value)
                    max_value = distances[key];
            }

            Pen istoPen = new Pen(Color.Orange, 2);

            foreach (int key in distances.Keys)
            {
                double currentInterValue = distances[key];
                double pct = (double)distances[key] / (double)max_value;
                g.DrawLine(istoPen,
                           new PointF(key, y - ((int)(pct * w))),
                           new PointF(key, y)
                );
            }

        }

    }
}